README.md
node-murmurhash
=======

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![Gittip][gittip-image]][gittip-url]
[![David deps][david-image]][david-url]
[![node version][node-image]][node-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/node-murmurhash.svg?style=flat-square
[npm-url]: https://npmjs.org/package/node-murmurhash
[travis-image]: https://img.shields.io/travis/node-modules/node-murmurhash.svg?style=flat-square
[travis-url]: https://travis-ci.org/node-modules/node-murmurhash
[coveralls-image]: https://img.shields.io/coveralls/node-modules/node-murmurhash.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/node-modules/node-murmurhash?branch=master
[gittip-image]: https://img.shields.io/gittip/fengmk2.svg?style=flat-square
[gittip-url]: https://www.gittip.com/fengmk2/
[david-image]: https://img.shields.io/david/node-modules/node-murmurhash.svg?style=flat-square
[david-url]: https://david-dm.org/node-modules/node-murmurhash
[node-image]: https://img.shields.io/badge/node.js-%3E=_0.10-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/
[download-image]: https://img.shields.io/npm/dm/node-murmurhash.svg?style=flat-square
[download-url]: https://npmjs.org/package/node-murmurhash

murmurhash V2, support utf8 Buffer.

## Install

```bash
$ npm install node-murmurhash --save
```

## Usage

```js
var murmurhash = require('node-murmurhash');

murmurhash('hello 中国', 97); // 1248731102
```

## Benchmark

```

node version: v4.2.1, date: Tue Dec 01 2015 14:06:01 GMT+0800 (CST)

Starting...
4 tests completed.

murmurhash(new Buffer('haha, this is key')) x 18,416,878 ops/sec ±0.70% (96 runs sampled)
murmurhash(new Buffer('快使用双节棍，嘿嘿嘿嘿。'))      x 12,709,085 ops/sec ±0.80% (93 runs sampled)
murmurhash('haha, this is key')             x    804,541 ops/sec ±24.21% (90 runs sampled)
murmurhash('快使用双节棍，嘿嘿嘿嘿。')                  x    797,345 ops/sec ±0.93% (92 runs sampled)
```

see also: [benchmark.js](test/benchmark.js)


## License

[MIT](LICENSE.txt)


History.md

2.1.0 / 2015-11-26
==================

 * feat: use Math.imul (@alsotang)

2.0.1 / 2015-07-21
==================

 * remove `& 0xff` when retrive byte (@alsotang)
 * simple benchmark. clean code.

2.0.0 / 2015-04-20
==================

 * test: add more boundary test cases
 * feat: pure javascript impl murmurhash v2. Thanks @zhangzifa

1.1.0 / 2015-02-04
==================

 * use node-pre-gyp-qn to prebuild binary package
 * upgrade nan to 1.5.2 support iojs and more node versions

1.0.1 / 2015-01-13
==================

 * Upgrade nan to 1.4.1 for support alinode/iojs

1.0.0 / 2014-09-25
==================

 * bump nan to 1.3.0 support node >= 0.11.13 now
 * add totoro ci tests
 * add missing line to murmur.cc
 * is hard to impl uint32 op in js

0.0.4 / 2014-03-22
==================

  * update deps and add jshint

0.0.3 / 2013-12-10
==================

  * remove node.h, because nan.h had include it

0.0.2 / 2013-12-10
==================

  * .node-gyp/0.11.8/src/node.h:176: error: ‘ssize_t’ does not name a type bug fix

0.0.1 / 2013-12-04
==================

  * add benchmark
  * add logo
  * init


test/benchmark.js
/*!
 * node-murmurhash - test/benchmark.js
 *
 * Copyright(c) fengmk2 <fengmk2@gmail.com> (http://fengmk2.com)
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var Benchmark = require('benchmark');
var benchmarks = require('beautify-benchmark');
var murmurhash = require('../');

var asciiBuffer = new Buffer('haha, this is key');
var utf8Buffer = new Buffer('快使用双节棍，嘿嘿嘿嘿。');
var asciiString = 'haha, this is key';
var utf8String = '快使用双节棍，嘿嘿嘿嘿。';

console.log('murmurhash should be 335538535: %s', murmurhash(asciiBuffer));
console.log('murmurhash utf8 should be 705333708: %s', murmurhash(utf8Buffer));
console.log('murmurhash with custom key should be 2639541842: %s', murmurhash(utf8Buffer, 12333));
console.log("murmurhash('hello 中国') should be 1248731102: %s", murmurhash('hello 中国'));

var suite = new Benchmark.Suite();

suite
.add("murmurhash(new Buffer('haha, this is key'))", function () {
  murmurhash(asciiBuffer);
})
.add("murmurhash(new Buffer('快使用双节棍，嘿嘿嘿嘿。'))", function () {
  murmurhash(utf8Buffer);
})
.add("murmurhash('haha, this is key')", function () {
  murmurhash(asciiString);
})
.add("murmurhash('快使用双节棍，嘿嘿嘿嘿。')", function () {
  murmurhash(utf8String);
})
.on('cycle', function(event) {
  benchmarks.add(event.target);
})
.on('start', function(event) {
  console.log('\n  node version: %s, date: %s\n  Starting...', process.version, Date());
})
.on('complete', function done() {
  benchmarks.log();
})
.run({ 'async': false });


test/longruning.js
/*!
 * node-murmurhash - test/longrunning.js
 *
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var murmurhash = require('../');

var i = 0;

function run() {
  i % 1000 === 0 && console.log(i, murmurhash(i));
  i++;
}

setInterval(run, 1);

// node --trace_gc test/longruning.js


test/murmurhash.test.js
/*!
 * node-murmurhash - test/murmurhash.test.js
 *
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com> (http://fengmk2.com)
 * Copyright(c) 2015 node-modules
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var should = require('should');
var murmurhash = require('../');

describe('murmurhash.test.js', function () {
  it('should return hash number value', function () {
    var v = murmurhash('foo');
    v.should.equal(2197288083);

    murmurhash('a').should.equal(3914231485);
    murmurhash('z').should.equal(2176583841);
    murmurhash('hello').should.equal(867496620);
    murmurhash('hello 中国').should.equal(1248731102);
    murmurhash('hello 中国', 0).should.equal(3760673533);
    murmurhash('hello 中国', 1).should.equal(540531333);

    var keybytes = new Buffer([
      0, 4, 111, 108, 95, 99, 51, 95, 117, 95, 105, 112, 95, -27, -116, -105, -26,
      -98, -127, -25, -69, -110, -25, -122, -108, -24, -98, -115, -28, -72, -109,
      -27, -115, -106, -27, -70, -105, 95, 112
    ]);
    // console.log(murmurhash(keybytes), parseInt(murmurhash(keybytes) % 1023, 10));
    murmurhash(keybytes).should.equal(3300337389);
    parseInt(murmurhash(keybytes) % 1023, 10).should.equal(261);

    var wantBytes = new Buffer([0, 24, 0, 4, 52, 51, 53, 54, 55, 52, 52, 53, 54]);
    murmurhash(wantBytes).should.equal(2351022266);
    parseInt(murmurhash(wantBytes) % 1023, 10).should.equal(494);
  });

  it('should support not string type', function () {
    murmurhash('1').should.equal(murmurhash(1));
  });

  it('should work with large buffer', function () {
    var buf = new Buffer([
      0x0,  0x1,  0x2,  0x3,  0x4,  0x5,  0x6,  0x7,  0x8,  0x9,  0xa,  0xb,
      0xc,  0xd,  0xe,  0xf,  0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17,
      0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 0x20, 0x21, 0x22, 0x23,
      0x24, 0x25, 0x26, 0x27, 0x28, 0x29, 0x2a, 0x2b, 0x2c, 0x2d, 0x2e, 0x2f,
      0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x3a, 0x3b,
      0x3c, 0x3d, 0x3e, 0x3f, 0x40, 0x41, 0x42, 0x43, 0x44, 0x45, 0x46, 0x47,
      0x48, 0x49, 0x4a, 0x4b, 0x4c, 0x4d, 0x4e, 0x4f, 0x50, 0x51, 0x52, 0x53,
      0x54, 0x55, 0x56, 0x57, 0x58, 0x59, 0x5a, 0x5b, 0x5c, 0x5d, 0x5e, 0x5f,
      0x60, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6a, 0x6b,
      0x6c, 0x6d, 0x6e, 0x6f, 0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77,
      0x78, 0x79, 0x7a, 0x7b, 0x7c, 0x7d, 0x7e, 0x7f,
      0xffffff80, 0xffffff81, 0xffffff82, 0xffffff83, 0xffffff84, 0xffffff85,
      0xffffff86, 0xffffff87, 0xffffff88, 0xffffff89, 0xffffff8a, 0xffffff8b,
      0xffffff8c, 0xffffff8d, 0xffffff8e, 0xffffff8f, 0xffffff90, 0xffffff91,
      0xffffff92, 0xffffff93, 0xffffff94, 0xffffff95, 0xffffff96, 0xffffff97,
      0xffffff98, 0xffffff99, 0xffffff9a, 0xffffff9b, 0xffffff9c, 0xffffff9d,
      0xffffff9e, 0xffffff9f, 0xffffffa0, 0xffffffa1, 0xffffffa2, 0xffffffa3,
      0xffffffa4, 0xffffffa5, 0xffffffa6, 0xffffffa7, 0xffffffa8, 0xffffffa9,
      0xffffffaa, 0xffffffab, 0xffffffac, 0xffffffad, 0xffffffae, 0xffffffaf,
      0xffffffb0, 0xffffffb1, 0xffffffb2, 0xffffffb3, 0xffffffb4, 0xffffffb5,
      0xffffffb6, 0xffffffb7, 0xffffffb8, 0xffffffb9, 0xffffffba, 0xffffffbb,
      0xffffffbc, 0xffffffbd, 0xffffffbe, 0xffffffbf, 0xffffffc0, 0xffffffc1,
      0xffffffc2, 0xffffffc3, 0xffffffc4, 0xffffffc5, 0xffffffc6, 0xffffffc7,
      0xffffffc8, 0xffffffc9, 0xffffffca, 0xffffffcb, 0xffffffcc, 0xffffffcd,
      0xffffffce, 0xffffffcf, 0xffffffd0, 0xffffffd1, 0xffffffd2, 0xffffffd3,
      0xffffffd4, 0xffffffd5, 0xffffffd6, 0xffffffd7, 0xffffffd8, 0xffffffd9,
      0xffffffda, 0xffffffdb, 0xffffffdc, 0xffffffdd, 0xffffffde, 0xffffffdf,
      0xffffffe0, 0xffffffe1, 0xffffffe2, 0xffffffe3, 0xffffffe4, 0xffffffe5,
      0xffffffe6, 0xffffffe7, 0xffffffe8, 0xffffffe9, 0xffffffea, 0xffffffeb,
      0xffffffec, 0xffffffed, 0xffffffee, 0xffffffef, 0xfffffff0, 0xfffffff1,
      0xfffffff2, 0xfffffff3, 0xfffffff4, 0xfffffff5, 0xfffffff6, 0xfffffff7,
      0xfffffff8, 0xfffffff9, 0xfffffffa, 0xfffffffb, 0xfffffffc, 0xfffffffd,
      0xfffffffe, 0xffffffff
    ]);
    // c++ => 2673451384
    murmurhash(buf).should.equal(2673451384);
  });

  it('should emtpy string got 3397915750', function () {
    murmurhash('').should.equal(3397915750);
    murmurhash(new Buffer('')).should.equal(3397915750);
    murmurhash(new Buffer([])).should.equal(3397915750);
  });

  it('should work with boundary cases: buffer mod length equal to 0,1,2,3', function () {
    // @zhangzifa:
    // 要覆盖到buffer的长度对4取模后余0/1/2/3个字节的情况
    // 并且这些尾数要分正负数情况

    var b = new Buffer([1, 2, 3, 4, 5, 6, 7, 8, 9]);
    murmurhash(b).should.equal(1543150919);
    var b1 = new Buffer([1, 2, 3, 4, 5, 6, 7, 8]);
    murmurhash(b1).should.equal(3881185275);
    var b2 = new Buffer([1, 2, 3, 4, 5, 6, 7]);
    murmurhash(b2).should.equal(782934980);
    var b3 = new Buffer([1, 2, 3, 4, 5, 6]);
    murmurhash(b3).should.equal(764621759);

    var nb = new Buffer([-1, -2, -3, -4, -5, -6, -7, -8, -9]);
    murmurhash(nb).should.equal(1119296559);
    var nb1 = new Buffer([-1, -2, -3, -4, -5, -6, -7, -8]);
    murmurhash(nb1).should.equal(3332328025);
    var nb2 = new Buffer([-1, -2, -3, -4, -5, -6, -7]);
    murmurhash(nb2).should.equal(1302582711);
    var nb3 = new Buffer([-1, -2, -3, -4, -5, -6]);
    murmurhash(nb3).should.equal(85049495);

    var bnb =  new Buffer([22, 145, 123, 149, 111, 156, 138, 238, 219]);
    murmurhash(bnb).should.equal(2935442895);
    var bnb1 = new Buffer([45, 222, 147, 158, 233, 156, 138, 238]);
    murmurhash(bnb1).should.equal(2003139132);
    var bnb2 = new Buffer([34, 255, 234, 168, 222, 145, 138]);
    murmurhash(bnb2).should.equal(2619540178);
    var bnb3 = new Buffer([88, 254, 222, 168, 134, 145]);
    murmurhash(bnb3).should.equal(3356056646);
  });
});


src/murmurhash.cc
#include "nan.h"
#include "murmurhash.h"

#define MURMURHASH_M 0x5bd1e995

NAN_METHOD(MurmurHash2) {
  NanScope();

  if (args.Length() < 2) {
    NanThrowTypeError("Wrong number of arguments");
    NanReturnUndefined();
  }

  char *key = node::Buffer::Data(args[0]->ToObject());
  int len = node::Buffer::Length(args[0]->ToObject());
  uint32_t seed = args[1]->NumberValue();
  uint32_t h = seed ^ len;
  int index = 0;

  while (len >= 4) {
    uint32_t k = (key[index] & 0xff) | ((key[index + 1] << 8) & 0xff00)
      | ((key[index + 2] << 16) & 0xff0000)
      | (key[index + 3] << 24);

    k *= MURMURHASH_M;
    k ^= (k >> 24);
    k *= MURMURHASH_M;
    h *= MURMURHASH_M;
    h ^= k;
    index += 4;
    len -= 4;
  }

  switch (len) {
    case 3:
      h ^= (key[index + 2] << 16);

    case 2:
      h ^= (key[index + 1] << 8);

    case 1:
      h ^= key[index];
      h *= MURMURHASH_M;
  }

  h ^= (h >> 13);
  h *= MURMURHASH_M;
  h ^= (h >> 15);
  NanReturnValue(NanNew<v8::Number>(h));
}


lib/murmur.js
/**!
 * node-modules - lib/murmur.js
 *
 * Copyright(c) node-modules and other contributors.
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <m@fengmk2.com> (http://fengmk2.com)
 *   zhangzifa <tonyzzf@163.com> (https://github.com/zhangzifa)
 */

'use strict';

/**
 * Module dependencies.
 */

module.exports = murmurhash2js;

var MURMURHASH_M = 0x5bd1e995;
var R16MURMURHASH_M = MURMURHASH_M & 0xffff;
var L16MURMURHASH_M = MURMURHASH_M >>> 16;

// @zhangzifa: c语言中，大于127的char, 在转成 int 时, 表示成0xffffffxx的形式
var NEGTIVE_MAP = [
  0x0,  0x1,  0x2,  0x3,  0x4,  0x5,  0x6,  0x7,  0x8,  0x9,  0xa,  0xb,
  0xc,  0xd,  0xe,  0xf,  0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17,
  0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f, 0x20, 0x21, 0x22, 0x23,
  0x24, 0x25, 0x26, 0x27, 0x28, 0x29, 0x2a, 0x2b, 0x2c, 0x2d, 0x2e, 0x2f,
  0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x3a, 0x3b,
  0x3c, 0x3d, 0x3e, 0x3f, 0x40, 0x41, 0x42, 0x43, 0x44, 0x45, 0x46, 0x47,
  0x48, 0x49, 0x4a, 0x4b, 0x4c, 0x4d, 0x4e, 0x4f, 0x50, 0x51, 0x52, 0x53,
  0x54, 0x55, 0x56, 0x57, 0x58, 0x59, 0x5a, 0x5b, 0x5c, 0x5d, 0x5e, 0x5f,
  0x60, 0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68, 0x69, 0x6a, 0x6b,
  0x6c, 0x6d, 0x6e, 0x6f, 0x70, 0x71, 0x72, 0x73, 0x74, 0x75, 0x76, 0x77,
  0x78, 0x79, 0x7a, 0x7b, 0x7c, 0x7d, 0x7e, 0x7f,
  0xffffff80, 0xffffff81, 0xffffff82, 0xffffff83, 0xffffff84, 0xffffff85,
  0xffffff86, 0xffffff87, 0xffffff88, 0xffffff89, 0xffffff8a, 0xffffff8b,
  0xffffff8c, 0xffffff8d, 0xffffff8e, 0xffffff8f, 0xffffff90, 0xffffff91,
  0xffffff92, 0xffffff93, 0xffffff94, 0xffffff95, 0xffffff96, 0xffffff97,
  0xffffff98, 0xffffff99, 0xffffff9a, 0xffffff9b, 0xffffff9c, 0xffffff9d,
  0xffffff9e, 0xffffff9f, 0xffffffa0, 0xffffffa1, 0xffffffa2, 0xffffffa3,
  0xffffffa4, 0xffffffa5, 0xffffffa6, 0xffffffa7, 0xffffffa8, 0xffffffa9,
  0xffffffaa, 0xffffffab, 0xffffffac, 0xffffffad, 0xffffffae, 0xffffffaf,
  0xffffffb0, 0xffffffb1, 0xffffffb2, 0xffffffb3, 0xffffffb4, 0xffffffb5,
  0xffffffb6, 0xffffffb7, 0xffffffb8, 0xffffffb9, 0xffffffba, 0xffffffbb,
  0xffffffbc, 0xffffffbd, 0xffffffbe, 0xffffffbf, 0xffffffc0, 0xffffffc1,
  0xffffffc2, 0xffffffc3, 0xffffffc4, 0xffffffc5, 0xffffffc6, 0xffffffc7,
  0xffffffc8, 0xffffffc9, 0xffffffca, 0xffffffcb, 0xffffffcc, 0xffffffcd,
  0xffffffce, 0xffffffcf, 0xffffffd0, 0xffffffd1, 0xffffffd2, 0xffffffd3,
  0xffffffd4, 0xffffffd5, 0xffffffd6, 0xffffffd7, 0xffffffd8, 0xffffffd9,
  0xffffffda, 0xffffffdb, 0xffffffdc, 0xffffffdd, 0xffffffde, 0xffffffdf,
  0xffffffe0, 0xffffffe1, 0xffffffe2, 0xffffffe3, 0xffffffe4, 0xffffffe5,
  0xffffffe6, 0xffffffe7, 0xffffffe8, 0xffffffe9, 0xffffffea, 0xffffffeb,
  0xffffffec, 0xffffffed, 0xffffffee, 0xffffffef, 0xfffffff0, 0xfffffff1,
  0xfffffff2, 0xfffffff3, 0xfffffff4, 0xfffffff5, 0xfffffff6, 0xfffffff7,
  0xfffffff8, 0xfffffff9, 0xfffffffa, 0xfffffffb, 0xfffffffc, 0xfffffffd,
  0xfffffffe, 0xffffffff
];

function murmurhash2js(key, seed) {
  var l = key.length;
  var h = seed ^ l;
  var i = 0;
  var k = 0;

  while (l >= 4) {
    k = (key[i] |
      (key[i + 1] << 8) |
      (key[i + 2] << 16) |
      (key[i + 3] << 24));

    // js 中，Number 能表示的最高精度整数是 (2 ** 53 - 1)，
    // 此处两个 32 位整数相乘时，有可能产生 64 位的结果，导致精度丢失。
    k = ((k * L16MURMURHASH_M) << 16) + k * R16MURMURHASH_M;

    k ^= k >>> 24;
    k = ((k * L16MURMURHASH_M) << 16) + k * R16MURMURHASH_M;

    h = (((h * L16MURMURHASH_M) << 16) + h * R16MURMURHASH_M) ^ k;

    l -= 4;
    i += 4;
  }

  switch (l) {
    case 3:
      h ^= NEGTIVE_MAP[key[i + 2]] << 16;
    case 2:
      h ^= NEGTIVE_MAP[key[i + 1]] << 8;
    case 1:
      h ^= NEGTIVE_MAP[key[i]];
      h = ((h * L16MURMURHASH_M) << 16) + h * R16MURMURHASH_M;
  }

  h ^= h >>> 13;
  h = ((h * L16MURMURHASH_M) << 16) + h * R16MURMURHASH_M;
  h ^= h >>> 15;

  return h >>> 0;
}
